experimental/oak_baremetal_channel/src/message.rs

```diff
@@ -0,0 +1,155 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use alloc::vec::Vec;
+
+pub type Length = u32;
+pub const LENGTH_SIZE: usize = 4;
+pub const LENGTH_OFFSET: usize = 0;
+static_assertions::assert_eq_size!([u8; LENGTH_SIZE], Length);
+
+pub type InvocationId = u32;
+pub const INVOCATION_ID_SIZE: usize = 4;
+pub const INVOCATION_ID_OFFSET: usize = 4;
+static_assertions::assert_eq_size!([u8; INVOCATION_ID_SIZE], InvocationId);
+
+pub type MethodId = u32;
+pub const METHOD_ID_SIZE: usize = 4;
+pub const METHOD_ID_OFFSET: usize = 8;
+static_assertions::assert_eq_size!([u8; METHOD_ID_SIZE], MethodId);
+
+pub type StatusCode = u32;
+pub const STATUS_CODE_SIZE: usize = 4;
+pub const STATUS_CODE_OFFSET: usize = 8;
+static_assertions::assert_eq_size!([u8; STATUS_CODE_SIZE], StatusCode);
+
+pub const PADDING_SIZE: usize = 4;
+
+pub const BODY_OFFSET: usize = 16;
+
+pub trait Message {
+    fn len(&self) -> usize;
+    fn encode(self) -> Vec<u8>;
+    fn decode(frames: Vec<u8>) -> Self;
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct RequestMessage {
+    pub invocation_id: InvocationId,
+    pub method_id: MethodId,
+    pub body: Vec<u8>,
+}
+
+impl Message for RequestMessage {
+    fn len(&self) -> usize {
+        LENGTH_SIZE + INVOCATION_ID_SIZE + METHOD_ID_SIZE + PADDING_SIZE + self.body.len()
+    }
+    fn encode(self) -> Vec<u8> {
+        let len = self.len();
+        let mut message_bytes: Vec<u8> = Vec::with_capacity(len);
+
+        let length: Length = len
+            .try_into()
+            .expect("could convert the message length usize to u32");
+
+        message_bytes.extend_from_slice(&length.to_le_bytes());
+        message_bytes.extend_from_slice(&self.invocation_id.to_le_bytes());
+        message_bytes.extend_from_slice(&self.method_id.to_le_bytes());
+        message_bytes.extend_from_slice(&[0; PADDING_SIZE]);
+        message_bytes.extend_from_slice(&self.body);
+
+        message_bytes
+    }
+    fn decode(mut encoded_message: Vec<u8>) -> Self {
```

Comment:
optional: perhaps add spacing between functions. The current case where there are spacing inside functions but no spacing between them makes it more difficult for me to quickly see the function boundaries.